otaclient/app/boot_control/_common.py

```diff
@@ -454,45 +470,49 @@ def _init_ota_status_files(self):
             return
 
-        # updating or rollbacking,
-        # with status==UPDATING or ROLLBACKING, we are expecting this is the first reboot
-        # during switching boot, and we should try finalizing the switch boot.
-        if _loaded_ota_status in [
+        # status except UPDATING and ROLLBACKING(like SUCCESS/FAILURE/ROLLBACK_FAILURE)
+        # are remained as it
+        if _loaded_ota_status not in [
             wrapper.StatusOta.UPDATING,
             wrapper.StatusOta.ROLLBACKING,
         ]:
-            # finalization failed or not in switching boot
-            if not (
-                self._is_switching_boot(self.active_slot)
-                and self.finalize_switching_boot()
-            ):
+            self._ota_status = _loaded_ota_status
+            return
+
+        # updating or rollbacking,
+        # if is_switching_boot, execute the injected finalize_switching_boot function from
+        #   boot controller, transit the ota_status according to the execution result.
+        # NOTE(20230614): for boot controller during multi-stage reboot(like rpi_boot),
+        #                 calling finalize_switching_boot might result in direct reboot,
+        #                 in such case, otaclient will terminate and ota_status will not be updated.
+        if self._is_switching_boot(self.active_slot):
+            if self.finalize_switching_boot():
+                self._ota_status = wrapper.StatusOta.SUCCESS
+                self._store_current_status(wrapper.StatusOta.SUCCESS)
+            else:
                 self._ota_status = (
                     wrapper.StatusOta.ROLLBACK_FAILURE
                     if _loaded_ota_status == wrapper.StatusOta.ROLLBACKING
                     else wrapper.StatusOta.FAILURE
                 )
                 self._store_current_status(self._ota_status)
-                _err_msg = "finalization failed on first reboot during switching boot"
-                logger.error(_err_msg)
-                raise BootControlInitError(_err_msg)
-
-            # finalizing switch boot successfully
-            self._ota_status = wrapper.StatusOta.SUCCESS
-            self._store_current_status(wrapper.StatusOta.SUCCESS)
-            return
-
-        # status except UPDATING and ROLLBACKING(like SUCCESS/FAILURE/ROLLBACK_FAILURE)
-        # are remained as it
-        else:
-            self._ota_status = _loaded_ota_status
-            # detect failed reboot, but only print error logging currently
-            if _loaded_slot_in_use != self.active_slot:
                 logger.error(
-                    f"boot into old slot {self.active_slot}, "
-                    f"but slot_in_use indicates it should boot into {_loaded_slot_in_use}, "
-                    "this might indicate a failed finalization at first reboot after update/rollback"
+                    "finalization failed on first reboot during switching boot"
                 )
```

Comment:
This line
```
raise BootControlInitError(_err_msg)
```
is removed, intentional?

Reply:
Yes, it is intentional, previously if boot init failed, the otaclient will just exit. With this change, it allows otaclient to keep alive and open the grpc server to report failure status.

Reply:
( And switch boot failed is not fatal error(comparing to boot control inits failed due to A/B partition layout incorrectly, etc), it should be counted as OTA failure, and reported via status API, instead of just crashing the otaclient I think.